forcecraft/aion
# enhancement(elixir): Handle user leaves

**Summary**

![giphy 3](https://user-images.githubusercontent.com/15965147/29741463-ac5b38a2-8a6d-11e7-8c81-bf791014097f.gif)

This PR introduces [Phoenix.Presence](https://hexdocs.pm/phoenix/Phoenix.Presence.html) to track  users who are present in a room. It also introduces changes mentioned in #40 - whenever you join a room and change your location, the change of game state should be registered.

The PR is not perfect, Presence is not sending out information when people are leaving the room empty. We will need to add a tool responsible for monitoring and handling that, but it shall be part of #85.

I have also moved out parts of the socket handling in Elm to *Socket.elm* because the *Update.elm* was getting really awful. 

**Related issues**

closes: #40 , #85, #89 

**Test plan**
Test it manually:
I:
1. Join a room using two accounts
2. Close browser tab responsible for one of the two players' game
3. See the updated list of players (user left)

![giphy 4](https://user-images.githubusercontent.com/15965147/29741464-b83597f8-8a6d-11e7-8404-7aa3023e6966.gif)

II:
1. Join a room using two accounts
2. Go back to the room lobby as one of the players
3. See the updates list of players (user left)

![giphy 5](https://user-images.githubusercontent.com/15965147/29741469-cc849f74-8a6d-11e7-9e37-292948905dc1.gif)
